File: practicum-java/src/main/java/ru/practicum/kafka/streams/util/WordMasker.java
package ru.practicum.kafka.streams.util;

import java.util.function.Predicate;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Маскирование запрещённых слов в тексте.
 *
 * <p>Вынесено из процессора отдельной чистой функцией: так логику можно проверить
 * обычным юнит-тестом, без топологии и брокера.
 */
public final class WordMasker {

    /** Чем заменяется запрещённое слово. */
    private static final char MASK_CHAR = '*';

    /**
     * Границы слова заданы явным классом символов, а не {@code \b}: в регулярных
     * выражениях Java по умолчанию {@code \w} — это только латиница, цифры
     * и подчёркивание, поэтому на кириллице границы срабатывали бы неверно.
     * Здесь словом считается последовательность букв и цифр любого алфавита.
     */
    private static final Pattern WORD_PATTERN = Pattern.compile("[\\p{L}\\p{N}]+", Pattern.UNICODE_CHARACTER_CLASS);

    private WordMasker() {
    }

    /**
     * Заменяет каждое слово, для которого {@code isBanned} вернул true, на звёздочки
     * той же длины. Пунктуация и пробелы сохраняются, длина текста не меняется.
     *
     * @param text     исходный текст сообщения
     * @param isBanned проверка слова по справочнику; слово приходит в нижнем регистре
     * @return текст с замаскированными словами
     */
    public static String mask(String text, Predicate<String> isBanned) {
        if (text == null || text.isEmpty()) {
            return text;
        }

        Matcher matcher = WORD_PATTERN.matcher(text);
        StringBuilder result = new StringBuilder(text.length());

        while (matcher.find()) {
            // Справочник хранится в нижнем регистре, поэтому сравнение
            // регистронезависимо: «Дурак» и «дурак» — одно и то же слово.
            String word = matcher.group();

            if (isBanned.test(word.toLowerCase())) {
                matcher.appendReplacement(result, Matcher.quoteReplacement(maskOf(word.length())));
            }
        }

        matcher.appendTail(result);
        return result.toString();
    }

    private static String maskOf(int length) {
        return String.valueOf(MASK_CHAR).repeat(length);
    }
}

File: practicum-dotnet/src/Practicum.Kafka.Streams/WordMasker.cs
using System.Text.RegularExpressions;

namespace Practicum.Kafka.Streams;

/// <summary>
/// Маскирование запрещённых слов в тексте.
///
/// Вынесено из процессора отдельной чистой функцией: так логику можно проверить
/// обычным тестом, без топологии и брокера.
/// </summary>
public static partial class WordMasker
{
    private const char MaskChar = '*';

    /// <summary>
    /// Словом считается последовательность букв и цифр любого алфавита.
    /// Явный класс символов вместо \b: в .NET \w по умолчанию включает Unicode,
    /// но опора на это неочевидна читателю, а здесь важно, что кириллица
    /// обрабатывается так же, как латиница.
    /// </summary>
    [GeneratedRegex(@"[\p{L}\p{N}]+")]
    private static partial Regex WordPattern();

    /// <summary>
    /// Заменяет каждое слово, для которого isBanned вернул true, на звёздочки
    /// той же длины. Пунктуация и пробелы сохраняются, длина текста не меняется.
    /// </summary>
    /// <param name="text">исходный текст сообщения</param>
    /// <param name="isBanned">проверка слова по справочнику; слово приходит в нижнем регистре</param>
    public static string? Mask(string? text, Func<string, bool> isBanned)
    {
        if (string.IsNullOrEmpty(text))
        {
            return text;
        }

        return WordPattern().Replace(text, match =>
        {
            // Справочник хранится в нижнем регистре, поэтому сравнение
            // регистронезависимо: «Дурак» и «дурак» — одно и то же слово.
            return isBanned(match.Value.ToLowerInvariant())
                ? new string(MaskChar, match.Length)
                : match.Value;
        });
    }
}
